contracts/HybridCustody.cdc

```diff
@@ -0,0 +1,529 @@
+
+    // Methods only accessible to the designated parent of a ProxyAccount
+    pub resource interface AccountPrivate {
+        pub fun getCapability(path: CapabilityPath, type: Type): Capability?
+        pub fun getPublicCapability(path: PublicPath, type: Type): Capability?
+
+        access(contract) fun redeemedCallback(_ addr: Address)
+    }
+
+    // Entry point for a parent to borrow its child account and obtain capabilities or
+    // perform other actions on the child account
+    pub resource interface ManagerPrivate {
+        pub fun borrowAccount(id: UInt64): &{AccountPrivate, AccountPublic}?
+        pub fun removeChildByAddress(addr: Address)
+        pub fun removeChild(id: UInt64)
+        // TODO: Owned account methods
+    }
+
+    // Functions anyone can call on a manager to get information about an account such as
+    // What child accounts it has
+    pub resource interface ManagerPublic {
+        pub fun borrowAccountPublic(id: UInt64): &{AccountPublic}?
+        // pub fun getChildAddresses(): [Address]
+    }
+
+    /*
+    Manager
+    A resource for an account which fills the Parent role of the Child-Parent account
+    management Model. A Manager can redeem or remove child accounts, and obtain any capabilities
+    exposed by the child account to them.
+    */
+    pub resource Manager: ManagerPrivate, ManagerPublic {
+        pub let accounts: {UInt64: Capability<&{AccountPrivate, AccountPublic}>}
+        pub let addressToAccountID: {Address: UInt64}
+
+        pub let ownedAccounts: {UInt64: Capability<&{Account, ChildAccountPrivate}>}
+        pub let addressToOwnedAccountID: {Address: UInt64}
+
+        pub fun addAccount(_ cap: Capability<&{AccountPrivate, AccountPublic}>) {
+            // Is there a scenario where you are shared the same address multiple times? Seems like overkill.
+            let acct = cap.borrow()
+                ?? panic("invalid account capability")
+
+            self.accounts[acct.uuid] = cap
+            self.addressToAccountID[cap.address] =  acct.uuid
+            
+            // TODO: emit account registered event
+
+            acct.redeemedCallback(self.owner!.address)
+        }
+
+        pub fun removeChildByAddress(addr: Address) {
+            let id = self.addressToAccountID.remove(key: addr)
+                ?? panic("no child account found with the given address")
+            self.removeChild(id: id)
+        }
+
+        pub fun removeChild(id: UInt64) {
+            let cap = self.accounts.remove(key: id) ?? panic("no account found with the given id")
+            self.addressToOwnedAccountID.remove(key: cap.address)
+
+            // TODO: emit event
+        }
+
+        pub fun addOwnedAccount(_ cap: Capability<&{Account, ChildAccountPrivate}>) {
+            let acct = cap.borrow()
+                ?? panic("cannot add invalid account")
+
+            self.ownedAccounts[acct.uuid] = cap
+            self.addressToOwnedAccountID[cap.address] = acct.uuid
```

Comment:
It will come down to whether we want to enforce that there are no duplicates for a parent managing two child accounts to the same address or not. If so, we would definitely want to add necessary pre-conditions like what you're suggesting